app/src/org/commcare/xml/AndroidBulkCaseXmlParser.java

```diff
@@ -0,0 +1,120 @@
+package org.commcare.xml;
+
+import android.net.ParseException;
+import android.net.Uri;
+import android.util.Pair;
+
+import net.sqlcipher.database.SQLiteDatabase;
+
+import org.commcare.CommCareApplication;
+import org.commcare.android.database.user.models.ACase;
+import org.commcare.cases.model.Case;
+import org.commcare.engine.references.JavaHttpReference;
+import org.commcare.interfaces.HttpRequestEndpoints;
+import org.commcare.logging.AndroidLogger;
+import org.commcare.models.database.SqlStorage;
+import org.commcare.models.database.user.models.AndroidCaseIndexTable;
+import org.commcare.models.database.user.models.EntityStorageCache;
+import org.commcare.utils.FileUtil;
+import org.commcare.utils.GlobalConstants;
+import org.commcare.xml.bulk.BulkProcessingCaseXmlParser;
+import org.javarosa.core.io.StreamsUtil;
+import org.javarosa.core.reference.InvalidReferenceException;
+import org.javarosa.core.reference.Reference;
+import org.javarosa.core.reference.ReferenceManager;
+import org.javarosa.core.services.Logger;
+import org.javarosa.core.services.storage.IStorageUtilityIndexed;
+import org.javarosa.core.util.PropertyUtils;
+import org.javarosa.xml.util.InvalidStructureException;
+import org.kxml2.io.KXmlParser;
+import org.xmlpull.v1.XmlPullParserException;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+
+/**
+ * A bulk processing parser for the android platform. Provides superior performance when
+ * processing high case loads during syncing and/or processing.
+ *
+ * @author ctsims
+ */
+public class AndroidBulkCaseXmlParser extends BulkProcessingCaseXmlParser {
+    private HttpRequestEndpoints generator;
+    private final EntityStorageCache mEntityCache;
+    private final AndroidCaseIndexTable mCaseIndexTable;
+    private final SqlStorage<ACase> storage;
+
+    public AndroidBulkCaseXmlParser(KXmlParser parser,
+                                         SqlStorage<ACase> storage,
+                                         HttpRequestEndpoints generator) {
+        this(parser, storage, new EntityStorageCache("case"), new AndroidCaseIndexTable(), generator);
+    }
+
+    public AndroidBulkCaseXmlParser(KXmlParser parser,
```

Comment:
It's used needed by tests. Still updating the framework to use them (the same way I did in the last one)